applications/wide_baseline_tracker_test.cpp

```diff
@@ -0,0 +1,169 @@
+ *  that the following conditions are met:
+ *
+ *  1. Redistributions of source code must retain the above copyright notice, this list of conditions and
+ *     the following disclaimer.
+ *
+ *  2. Redistributions in binary form must reproduce the above copyright notice, this list of conditions and
+ *     the following disclaimer in the documentation and/or other materials provided with the distribution.
+ * 
+ *  3. Neither the name of the copyright holder nor the names of its contributors may be used to endorse or
+ *     promote products derived from this software without specific prior written permission.
+ * 
+ *  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, *  INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ *  DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ *  SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+ *  SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
+ *  WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE
+ *  USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ *
+ *  Author:
+ *
+ *  Luiz Felipe Maciel Correia
+ *  Marcos Henrique Marcone
+ *  Bruno Silva
+ */
+
+#include <cstdio>
+#include <cstdlib>
+#include <opencv2/core/core.hpp>
+#include <opencv2/highgui/highgui.hpp>
+#include <opencv2/imgproc/imgproc.hpp>
+
+#include <rgbd_loader.h>
+#include <wide_baseline_tracker.h>
+#include <config_loader.h>
+#include <event_logger.h>
+#include <common_types.h>
+
+using namespace std;
+using namespace cv;
+
+void draw_last_track(Mat& img, const vector<Point2f> prev_pts, const vector<Point2f> curr_pts, bool is_kf);
+void draw_tracks(Mat &img, const vector<Tracklet> tracklets);
+
+/**
+ * This program shows the use of tracking by detection algorithm.
+ * @param .yml config. file (from which index_file is used)
+ */
+int main(int argc, char **argv)
+{
+	EventLogger& logger = EventLogger::getInstance();
+	logger.setVerbosityLevel(pcl::console::L_INFO);
+	
+	RGBDLoader loader;
+	string index_file;
+	int log_stats;
+	Mat frame, depth, current_frame, previous_frame;
+	Ptr<cv::FeatureDetector> feature_detector;
+	Ptr<cv::DescriptorExtractor> descriptor_extractor;
+	Ptr<cv::DescriptorMatcher> descriptor_matcher;
+	
+	if (argc != 2)
+	{
+		logger.print(pcl::console::L_INFO, "[wide_baseline_tracker_test.cpp] Usage: %s <path/to/config_file.yaml>\n", argv[0]);
+		exit(0);
+	}
+
+	ConfigLoader param_loader(argv[1]);
+	param_loader.checkAndGetString("index_file", index_file);
+	param_loader.checkAndGetInt("log_stats", log_stats);
+	param_loader.checkAndGetFeatureDetector("feature_detector", feature_detector);
```

Comment:
The relation between the WideBaselineTracker and the feature detector/extractor and matcher should be a composition instead of an aggregation.

Change the ConfigLoader to return strings informing which detector/extractor/matcher (instead of returning Ptr) and change the WideBaselineTracker constructor to accept strings instead of Ptr too.

Reply:
The changes that resolve this case are in the commit 429d944b5468eff2331c6fa37d37ee593979f774